lib/Twig/Extension/Escaper.php

```diff
@@ -11,16 +11,95 @@
 
 final class Twig_Extension_Escaper extends Twig_Extension
 {
+    private $environment;
+    private $core;
     private $defaultStrategy;
+    private $escapersSafe = array(
+        'html' => array('html'),
+        'js' => array('js'),
+        'css' => array('css'),
+        'html_attr' => array('html', 'html_attr'),
+        'url' => array('url'),
+    );
 
     /**
      * @param string|false|callable $defaultStrategy An escaping strategy
+     * @param Twig_Environment      $env             Environment this extension is added to. Used temporarily to deprecate Twig_Extension_Core::setEscaper()/getEscapers() instead of removing them without notice.
      *
      * @see setDefaultStrategy()
      */
-    public function __construct($defaultStrategy = 'html')
+    public function __construct($defaultStrategy = 'html', Twig_Environment $env = null)
     {
         $this->setDefaultStrategy($defaultStrategy);
+        $this->environment = $env;
```

Comment:
I've moved the signature code to methods on Twig_Extension. It takes a bit more work to check if the signature needs update (loops through extensions and checks each individually), but it gets rid of the circular reference for the signature. (Committed separately from the other changes in case you'd rather keep the circular reference.)

There's still a dependency on Environment in Escaper, but it's to pull in Core for `setEscaper()` and `getEscaper()`. As far as I can tell, there's no way around this, except to wait to push this through until 3.x (which would allow adding the new methods on Twig_Extension to Twig_ExtensionInterface).